src/lib/libwasi.js

```diff
@@ -84,27 +84,31 @@ var WasiLibrary = {
     return getEnvStrings.strings;
   },
 
-  environ_sizes_get__deps: ['$getEnvStrings'],
+  environ_sizes_get__deps: ['$getEnvStrings', '$lengthBytesUTF8'],
   environ_sizes_get__nothrow: true,
   environ_sizes_get: (penviron_count, penviron_buf_size) => {
     var strings = getEnvStrings();
     {{{ makeSetValue('penviron_count', 0, 'strings.length', SIZE_TYPE) }}};
     var bufSize = 0;
-    strings.forEach((string) => bufSize += string.length + 1);
+    for (var string of strings) {
+      bufSize += lengthBytesUTF8(string) + 1;
+    }
     {{{ makeSetValue('penviron_buf_size', 0, 'bufSize', SIZE_TYPE) }}};
     return 0;
   },
 
-  environ_get__deps: ['$getEnvStrings', '$stringToAscii'],
+  environ_get__deps: ['$getEnvStrings', '$stringToUTF8'],
   environ_get__nothrow: true,
   environ_get: (__environ, environ_buf) => {
     var bufSize = 0;
-    getEnvStrings().forEach((string, i) => {
+    var envp = 0;
+    for (var string of getEnvStrings()) {
       var ptr = environ_buf + bufSize;
-      {{{ makeSetValue('__environ', `i*${POINTER_SIZE}`, 'ptr', POINTER_TYPE) }}};
-      stringToAscii(string, ptr);
+      {{{ makeSetValue('__environ', 'envp', 'ptr', POINTER_TYPE) }}};
+      stringToUTF8(string, ptr, Infinity);
       bufSize += string.length + 1;
```

Comment:
Isn't this using the length of the JS string, not the encoded C string? In non-ascii those can diverge iiuc

Reply:
Oooops good catch.